google-cloud-spanner/lib/google/cloud/spanner/pool.rb

```diff
@@ -154,32 +163,91 @@ def reset
 
         def close
           @thread.kill if @thread
-          @all_sessions.each(&:delete_session)
-          @all_sessions = []
-          @session_queue = []
-          @transaction_queue = []
+          # Unblock all waiting threads
+          @closed = true
+          @resource.broadcast
+          # Delete all sessions
+          @mutex.synchronize do
+            @all_sessions.map do |s|
+              Thread.new do
+                s.release!
+              end
+            end.map(&:join)
+            @all_sessions = []
+            @session_queue = []
+            @transaction_queue = []
+          end
 
           true
         end
 
+        def keepalive_or_release!
+          to_keepalive = []
+          to_release = []
+
+          @mutex.synchronize do
+            available_count = session_queue.count + transaction_queue.count
```

Comment:
If there are more idle sessions than the minimum number of sessions configured, then the difference will be released instead of kept alive. So if there are 15 idle sessions (haven't been used in 30 minutes), and the minimum number of sessions is 10, 10 will be kept alive and 5 will be released.